src/analysis.py

```diff
@@ -198,79 +215,129 @@ def train_models(train_df, models, param_grid=None, output_dir=""):
-            random_search = RandomizedSearchCV(random_forest_pipeline, 
-                                               param_distributions=param_grid, 
-                                               cv=5, 
-                                               n_jobs=-1, 
-                                               n_iter=20, 
-                                               return_train_score=True)
-            random_search.fit(X_train, y_train)
-            if output_dir:
-                pickle.dump(random_search.best_estimator_, 
-                open(output_dir + "/models/optimized_model.pkl", 'wb'))
-
-        results = pd.DataFrame(random_search.cv_results_).set_index("rank_test_score").sort_index()
-        results.reset_index(inplace=True)
-        results = results.rename(columns = {'index':'Ranked Test Scores'})
-        return results
-
-
     for key, value in models.items():
         pipeline = Pipeline(
                 steps=[
                     ("preprocessor", preprocessor),
                     (key, value)
                 ]
             )
-        scores = cross_validate(pipeline, X_train, y_train, return_train_score=True, n_jobs=-1, verbose=1)
-        store_cross_val_results(key, scores, results_df)
+        results_df[key] = store_cross_val_results(pipeline, X_train, y_train)
 
     results = pd.DataFrame(results_df)
     results.reset_index(inplace=True)
     results = results.rename(columns = {'index':'Scores'})
     return results
 
+def optimize_model(train_df, param_grid=None, output_dir="", fileprefix=""):
+    """
+    Processes the data, performs hyperparameter optimization on the model, and returns that model
+
+    Parameters
+    ----------
+    train_df : pd.Dataframe
+        Dataframe to be used in model
+
+    param_grid : dict
+        hyperparameters for model
+
+    ouput_dir : string
+        The directory where the output of the model will be generated
+
+    fileprefix : string
+        Prefix to be added to the filename of the output
+
+    Returns
+    ----------
+        pandas.Dataframe
+
+    """
+    X_train = train_df.drop(columns=['align'])
+    y_train = train_df['align']
+
+    numeric_features = ['appearances', 'year', 'name_len', 'appear_per_yr']
+    categorical_features = ['id', 'eye', 'hair', 'publisher', 'sex']
+    drop_features = ['name', 'gsm', 'first_appearance', 'first_name', 'last_name']
+    binary_features = ['is_common', 'last_name']
 
-def save_img(data_frame, output_folder, file_name):
-    data_frame.to_pickle(output_folder + "/tables/" + file_name + ".pkl")
+    numeric_transformer = Pipeline(
+        steps=[
+            ("imputer", SimpleImputer(strategy='median')),
+            ("scaler", StandardScaler()),
+        ]
+    )
+
+    categorical_transformer = Pipeline(
+        steps=[
+            ("imputer", SimpleImputer(strategy='most_frequent')),
+            ("onehot", OneHotEncoder(handle_unknown='ignore'))
+        ]
+    )
+
+    binary_transformer = Pipeline(
```

Comment:
If you decide to adopt the code above for binary_transformer, you may want to update this as well.

Reply:
See above comment!